packages/emails/lib/generateIcsString.ts

```diff
@@ -0,0 +1,142 @@
+import type { DateArray, ParticipationStatus, ParticipationRole, EventStatus } from "ics";
+import { createEvent } from "ics";
+import type { TFunction } from "next-i18next";
+import { RRule } from "rrule";
+
+import dayjs from "@calcom/dayjs";
+import { getRichDescription } from "@calcom/lib/CalEventParser";
+import { getWhen } from "@calcom/lib/CalEventParser";
+import type { CalendarEvent, Person } from "@calcom/types/Calendar";
+
+export enum BookingAction {
+  Create = "create",
+  Cancel = "cancel",
+  Reschedule = "reschedule",
+  RequestReschedule = "request_reschedule",
+  LocationChange = "location_change",
+}
+
+const generateIcsString = ({
+  event,
+  t,
+  role,
+  bookingAction,
+}: {
+  event: CalendarEvent;
+  t: TFunction;
+  role: "attendee" | "organizer";
+  bookingAction: BookingAction;
+}) => {
+  let title = "",
+    subtitle = "",
+    status: EventStatus = "TENTATIVE";
+  // Taking care of recurrence rule
+  let recurrenceRule: string | undefined = undefined;
+  const partstat: ParticipationStatus = "ACCEPTED";
+  const icsRole: ParticipationRole = "REQ-PARTICIPANT";
+  if (event.recurringEvent?.count) {
+    // ics appends "RRULE:" already, so removing it from RRule generated string
+    recurrenceRule = new RRule(event.recurringEvent).toString().replace("RRULE:", "");
+  }
+
+  switch (bookingAction) {
+    case BookingAction.Create:
+      if (role === "organizer") {
+        title = event.recurringEvent?.count ? "new_event_scheduled_recurring" : "new_event_scheduled";
+      } else if (role === "attendee") {
+        title = event.recurringEvent?.count
+          ? "your_event_has_been_scheduled_recurring"
+          : "your_event_has_been_scheduled";
+      }
+      status = "CONFIRMED";
+      break;
+    case BookingAction.Cancel:
+      title = "event_request_cancelled";
+      status = "CANCELLED";
+      break;
+    case BookingAction.Reschedule:
+      title = "event_has_been_rescheduled";
+      status = "CONFIRMED";
+      break;
+    case BookingAction.RequestReschedule:
+      if (role === "organizer") {
+        title = t("request_reschedule_title_organizer", {
+          attendee: event.attendees[0].name,
+        });
+        subtitle = t("request_reschedule_subtitle_organizer", {
+          attendee: event.attendees[0].name,
+        });
+      } else if (role === "attendee") {
+        title = "request_reschedule_booking";
+        subtitle = t("request_reschedule_subtitle", {
+          organizer: event.organizer.name,
+        });
+      }
+      status = "CANCELLED";
+      break;
+    case BookingAction.LocationChange:
+      title = "event_location_changed";
+      break;
+  }
+
+  const getTextBody = (title = "", subtitle = "emailed_you_and_any_other_attendees"): string => {
+    let body: string;
+    if (BookingAction.RequestReschedule && role === "attendee") {
```

Comment:
It shouldn't be it's responsibility to manage content based on action. We should just pass the content.
If we need to pass a function then we can do that too. We could just let `generateIcsString` accept a fn called getTextBody.

Reply:
With `getTextBody` I remember making changes to emails and having to make the change in every email class. Requesting reschedule is the only email where it differs. Could we have a default `getTextBody` we can use and if a function is passed then use that to generate the text body?

Reply:
Yeah maybe we can define a utility `getTextBody` and we can pass it from all places to `generateIcsString` except the rescheduling case which would have different `getTextBody`. In this way `generateIcsString` doesn't care about rescheduling and it is just concerned about generating an consistent ics string.